aws/resource_aws_dx_connection.go

```diff
@@ -14,79 +16,142 @@ func resourceAwsDxConnection() *schema.Resource {
+			},
+			"name": {
 				Type:     schema.TypeString,
 				Required: true,
 				ForceNew: true,
 			},
-			"bandwidth": &schema.Schema{
+			"bandwidth": {
 				Type:         schema.TypeString,
 				Required:     true,
 				ForceNew:     true,
 				ValidateFunc: validateDxConnectionBandWidth,
 			},
-			"location": &schema.Schema{
+			"location": {
 				Type:     schema.TypeString,
 				Required: true,
 				ForceNew: true,
 			},
+			"tags": tagsSchema(),
 		},
 	}
 }
 
 func resourceAwsDxConnectionCreate(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).dxconn
 
-	input := &directconnect.CreateConnectionInput{
+	req := &directconnect.CreateConnectionInput{
 		Bandwidth:      aws.String(d.Get("bandwidth").(string)),
 		ConnectionName: aws.String(d.Get("name").(string)),
 		Location:       aws.String(d.Get("location").(string)),
 	}
-	resp, err := conn.CreateConnection(input)
+
+	log.Printf("[DEBUG] Creating Direct Connect connection: %#v", req)
+	resp, err := conn.CreateConnection(req)
 	if err != nil {
 		return err
 	}
-	d.SetId(*resp.ConnectionId)
-	return resourceAwsDxConnectionRead(d, meta)
+
+	d.SetId(aws.StringValue(resp.ConnectionId))
+	return resourceAwsDxConnectionUpdate(d, meta)
 }
 
 func resourceAwsDxConnectionRead(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).dxconn
 
-	connectionId := d.Id()
-	input := &directconnect.DescribeConnectionsInput{
-		ConnectionId: aws.String(connectionId),
-	}
-	resp, err := conn.DescribeConnections(input)
+	resp, err := conn.DescribeConnections(&directconnect.DescribeConnectionsInput{
+		ConnectionId: aws.String(d.Id()),
+	})
 	if err != nil {
+		if isNoSuchDxConnectionErr(err) {
+			log.Printf("[WARN] Direct Connect connection (%s) not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
 		return err
 	}
+
 	if len(resp.Connections) < 1 {
+		log.Printf("[WARN] Direct Connect connection (%s) not found, removing from state", d.Id())
 		d.SetId("")
 		return nil
 	}
 	if len(resp.Connections) != 1 {
-		return fmt.Errorf("[ERROR] Number of DX Connection (%s) isn't one, got %d", connectionId, len(resp.Connections))
+		return fmt.Errorf("[ERROR] Number of Direct Connect connections (%s) isn't one, got %d", d.Id(), len(resp.Connections))
 	}
-	if d.Id() != *resp.Connections[0].ConnectionId {
-		return fmt.Errorf("[ERROR] DX Connection (%s) not found", connectionId)
+	connection := resp.Connections[0]
+	if d.Id() != aws.StringValue(connection.ConnectionId) {
+		return fmt.Errorf("[ERROR] Direct Connect connection (%s) not found", d.Id())
+	}
+	terminalStates := map[string]bool{
+		directconnect.ConnectionStateDeleted:  true,
+		directconnect.ConnectionStateDeleting: true,
+		directconnect.ConnectionStateRejected: true,
+	}
+	if _, ok := terminalStates[aws.StringValue(connection.ConnectionState)]; ok {
+		log.Printf("[WARN] Direct Connect connection (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
```

Comment:
I'd be ok with removing the resource from state if it's in `ConnectionStateDeleted`, but I think we should keep it in the state if it's still `ConnectionStateDeleting` or `ConnectionStateRejected`. The deletion may not finish successfully and rejection != deleted resource, AFAIK?

Reply:
Yes, agreed. I'll fix.